executor/src/witgen/jit/processor.rs

```diff
@@ -0,0 +1,406 @@
+                    continue;
+                }
+                let result = witgen.process_identity(can_process.clone(), id, *row_offset);
+                progress |= result.progress;
+                if result.complete {
+                    complete.insert((id.id(), *row_offset));
+                }
+            }
+        }
+    }
+
+    /// If any machine call could not be completed, that's bad because machine calls typically have side effects.
+    /// So, the underlying lookup / permutation / bus argument likely does not hold.
+    /// This function checks that all machine calls are complete, at least for a window of <block_size> rows.
+    /// It returns the list of incomplete calls, if any.
+    fn incomplete_machine_calls(&self, complete: &HashSet<(u64, i32)>) -> Vec<(&Identity<T>, i32)> {
+        self.identities
+            .iter()
+            .map(|(id, _)| id)
+            .filter(|id| is_machine_call(id))
+            .unique()
+            .flat_map(|&call| {
+                let rows = self.rows_for_identity(call);
+                let complete_rows = rows
+                    .iter()
+                    .filter(|&&row| complete.contains(&(call.id(), row)))
+                    .collect::<Vec<_>>();
+                // Because we process rows -1..block_size+1, it is fine to have two incomplete machine calls,
+                // as long as <block_size> consecutive rows are complete.
+                if complete_rows.len() >= self.block_size {
+                    let (min, max) = complete_rows.iter().minmax().into_option().unwrap();
+                    // TODO instead of checking for consequitive rows, we could also check
+                    // that they "fit" the next block.
+                    // TODO actually I think that we should not allow more than block size
+                    // completed calls.
+                    let is_consecutive = *max - *min == complete_rows.len() as i32 - 1;
+                    if is_consecutive {
+                        return vec![];
+                    }
+                }
+                rows.iter()
+                    .filter(|&row| !complete.contains(&(call.id(), *row)))
+                    .map(|row| (call, *row))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Returns the list of rows the given identity is processed on.
+    fn rows_for_identity(&self, identity: &Identity<T>) -> Vec<i32> {
+        self.identities
+            .iter()
+            .filter_map(move |(id, row_offset)| {
+                if *id == identity {
+                    Some(*row_offset)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    /// After solving, the known cells should be such that we can stack different blocks.
+    /// TODO the same is actually true for machine calls.
+    fn check_block_shape(
```

Comment:
Why put this here, if it is not run when called from the single step processor?

Reply:
Because the block_machine_processor does not have access to witgen any more

Reply:
Also it needs to be run in each branch.

Reply:
Right. But that means that the machine could have a different shape in each branch!

I think the proper way to handle this would be to let the processor do it's thing and then look at the referenced cells of the generated effects.

Reply:
-> https://github.com/powdr-labs/powdr/issues/2327